### PR TITLE
add async commit

### PR DESCRIPTION
**Pull request by CharLotteiu**

### CODE REVIEW

**Comment by CharLotteiu** [comment]
@CaitinChen PTAL

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+image: /images/blog/async-commit-the-accelerator-for-transaction-commit.jpg
+---
+
+**Author:** [Yilin Chen](https://github.com/sticnarf) (Software Engineer at PingCAP, TiKV Commiter)
+
+**Editor:** Tom Dewan; [Charlotte Liu](https://github.com/CharLotteiu)
+
+![Async Commit](media/async-commit-the-accelerator-for-transaction-commit.jpg)
+
+TiDB is an open source, distributed, scale-out MySQL alternative database that supports Hybrid Transactional and Analytical Processing (HTAP) workloads. It provides native support for distributed transactions.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:17]:
```suggestion
[TiDB](https://docs.pingcap.com/tidb/stable) is an open source, distributed, scale-out MySQL alternative database that supports Hybrid Transactional and Analytical Processing (HTAP) workloads. It provides native support for distributed transactions.
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+As TiDB's developers, we have been working on implementing low-latency distributed transactions. Recently, TiDB 5.0 takes a big step to greatly reduce the latency of transaction commits—introducing the **Async Commit** feature.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:19]:
```suggestion
As TiDB's developers, we have been working on implementing low-latency distributed transactions. Recently, [TiDB 5.0](https://docs.pingcap.com/tidb/stable/release-5.0.0) takes a big step to greatly reduce the latency of transaction commits—introducing the **Async Commit** feature.
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+date: 2021-05-17
+summary: As a major new feature introduced in TiDB 5.0, Async Commit can greatly reduce the latency of transaction commits. This post describes Async Commit's design ideas and its key implementation details.
+tags: ['TiDB 5.0', 'Transaction']
+categories: ['Engineering']
+image: /images/blog/async-commit-the-accelerator-for-transaction-commit.jpg
+---
+
+**Author:** [Yilin Chen](https://github.com/sticnarf) (Software Engineer at PingCAP, TiKV Commiter)
+
+**Editor:** Tom Dewan; [Charlotte Liu](https://github.com/CharLotteiu)
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:13]:
```suggestion
**Editors:** Tom Dewan, [Charlotte Liu](https://github.com/CharLotteiu)
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+<div class="caption-center"> The commit process before Async Commit is introduced </div>
+
+After the user sends a COMMIT statement to TiDB, TiDB has to go through at least the following steps before returning the commit result to the user:
+
+1. Concurrently prewrites all keys.
+2. Gets a timestamp from Placement Driver (PD) as the commit timestamp (Commit TS).
+3. Commits the primary key.
+
+The critical path for the entire commit process includes at least two round trips between TiDB and TiKV. Only secondary keys are committed asynchronously in the background.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:39]:
```suggestion
The critical path for the entire commit process includes at least two round trips between [TiDB](https://docs.pingcap.com/tidb/stable/tidb-architecture#tidb-server) and [TiKV](https://docs.pingcap.com/tidb/stable/tidb-architecture#tikv-server). Only secondary keys are committed asynchronously in the background.
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+<div class="caption-center"> The secondary keys store pointers to the primary key </div>
+
+As for an Async Commit transaction, we need to know the states of all keys to determine the transaction state; that is, we must be able to query every key of the transaction from any key in the transaction. Therefore, we also store the pointers to each secondary key in the entry of the primary key:
+
+![Pointers to the secondary keys](media/pointers-to-the-secondary-keys.png)
+
+<div class="caption-center"> The primary key stores pointers to the secondary keys </div>
+
+The primary key actually stores a list of all secondary keys. Obviously, if a transaction contains a large number of keys, we cannot store them all on the primary key. So an Async Commit transaction cannot be too large. Currently, we only use Async Commit for transactions that contain up to 256 keys, and the total size of all keys must be no more than 4096 bytes.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:74]:
```suggestion
The primary key actually stores a list of all secondary keys. Obviously, if a transaction contains a large number of keys, we cannot store them all on the primary key. So an Async Commit transaction cannot be too large. Currently, we only use Async Commit for transactions that contain up to 256 keys, and the total size of all keys must be no more than 4,096 bytes.
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+In summary, the Min Commit TS of each key takes the maximum value between the Max TS + 1 that TiDB gets when prewriting, and the timestamp obtained from the PD before the prewriting process. The Commit TS of a transaction is the maximum Min Commit TS of all keys. They together guarantee both snapshot isolation and linearizability for Async Commit transactions.
+
+### One-phase commit (1PC)
+
+If a transaction only updates a non-index column of a record, or inserts a record without a secondary index, it only involves a single region. In this scenario, is it possible to complete the transaction commit in only one phase without using the distributed transaction protocol? Well, this is certainly possible, but the difficulty lies in how to determine the Commit TS for a one-phase commit transaction.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:148]:
```suggestion
If a transaction only updates a non-index column of a record, or inserts a record without a secondary index, it only involves a single [Region](https://docs.pingcap.com/tidb/stable/glossary#regionpeerraft-group). In this scenario, is it possible to complete the transaction commit in only one phase without using the distributed transaction protocol? Well, this is certainly possible, but the difficulty lies in how to determine the Commit TS for a one-phase commit transaction.
```

Comment (CaitinChen) [reply async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:148]: "Region" is a technical term for TiDB. Please update all occurrences. Thanks.

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+As it is not very usual for such a special scenario to occur, we still offer users a way to skip getting the Min Commit TS. Transactions opened with `START TRANSACTION WITH CAUSAL CONSISTENCY ONLY` do not get the Min Commit TS when committing. If your scenario does not involve controlling the commit order of two concurrent transactions outside the database, you can try to lower the consistency level to save one operation of TiDB getting the timestamp from the PD TSO, and thus reduce the time.
+
+## Performance enhancement
+
+Async Commit moves the transaction completion point to the end of prewrite, making the commit of the primary key asynchronous. The more time consuming the primary key commit step is in the transaction, the more significant the improvement brought by Async Commit will be. Smaller transactions with fewer interactions can often gain large improvement through Async Commit.
+
+In the sysbench oltp_update_index scenario, a transaction only writes two keys, row record, and index. It is also an auto-commit transaction with no additional interaction, so theoretically Async Commit can significantly reduce its latency.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:278]:
```suggestion
In the sysbench `oltp_update_index` scenario, a transaction only writes two keys, row record, and index. It is also an auto-commit transaction with no additional interaction, so theoretically Async Commit can significantly reduce its latency.
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+Our tests also prove this. As shown in the below chart, when we enable Async Commit and test sysbench oltp_update_index at a fixed 2,000 TPS, the average latency is reduced by 42% and the 99th percentile latency is reduced by 32%.
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:280]:
```suggestion
Our tests also prove this. As shown in the below chart, when we enable Async Commit and test sysbench `oltp_update_index` at a fixed 2,000 TPS, the average latency is reduced by 42% and the 99th percentile latency is reduced by 32%.
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+The following diagram shows the commit process of an Async Commit transaction. You can see that the original part of getting the Commit TS is missing. Instead, TiDB gets a timestamp from PD as the Min Commit TS before the prewrites. The reasons for the change will be explained in the following sections.
+
+![The commit process of an Async Commit transaction](media/commit-process-of-an-async-commit-transaction.png)
+
+<div class="caption-center"> The commit process of an Async Commit transaction </div>
+
+To reach the goal of advancing the time when the transaction state is determined, there are two main issues that need to be addressed:
+
+* How to determine if all keys have been prewritten?
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:57]:
```suggestion
* How to determine if all keys have been prewritten
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+* How to determine the Commit TS of a transaction?
```
Comment (CaitinChen) [inline async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md:58]:
```suggestion
* How to determine the Commit TS of a transaction
```

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md (hunk 6 above)
Comment (CharLotteiu) [reply async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md]: Updated in https://github.com/pingcap/blog/pull/581/commits/54d3197c266d03f140d0abd3c62b1c4261eda1a3.

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md (hunk 9 above)
Comment (CharLotteiu) [reply async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md]: Updated in https://github.com/pingcap/blog/pull/581/commits/54d3197c266d03f140d0abd3c62b1c4261eda1a3.

File: async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md (hunk 10 above)
Comment (CharLotteiu) [reply async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md]: Updated in https://github.com/pingcap/blog/pull/581/commits/54d3197c266d03f140d0abd3c62b1c4261eda1a3.

**Review by CaitinChen — APPROVED** [review]
LGTM

**Pull request merged**